### PR TITLE
New REST API Endpoints for the block editor

### LINKED ISSUE #78: Expose functionality through WP REST API
In order to make the plugin work with a native Block Editor UI as described in #45 all the actions need to be accessible through the REST API.

That means:
 - [ ] Getting all the connections that are available for the current post type
 - [ ] Getting the posts that are currently connected to an entity
 - [ ] Setting the posts that are currently connected to an entity
 - [ ] Adding a new connected post
 - [ ] Removing a connected post

### PR DESCRIPTION
**Pull request by s3rgiosan**
> [!WARNING]
> This PR is based on top of #96

### Description of the Change

This PR adds REST API endpoints for searching, retrieving, and managing post relationships, improving integration with the Block Editor and other WordPress components. The implementation leverages the helper functions introduced in #96 to provide a clean, modular API.

**V1 Endpoints:**
- `POST /content-connect/v1/search` - Search for posts and users with nonce-based authentication, supporting pagination and filtering by post type or relationship.

**V2 Endpoints:**
- `GET /content-connect/v2/relationships` - Retrieves all relationships by type, with optional filtering by key, post type, or from/to post types.
- `GET /content-connect/v2/post/{post_id}/relationships` - Retrieve all relationships associated with a post, with support for filtering by relationship type, post type, and context (view/embed).
- `GET /content-connect/v2/post/{post_id}/related` - Retrieve entities related to a post for a specific relationship, with pagination, ordering, and filtering support.
- `POST /content-connect/v2/post/{post_id}/related` - Update the complete list of related entities for a post, preserving sort order for sortable relationships.
- `PUT /content-connect/v2/post/{post_id}/related` - Add a single related entity to a post.
- `DELETE /content-connect/v2/post/{post_id}/related` - Remove a related entity from a post.

**REST API Link Integration:**
- Automatically adds `content-connect:relationships` and `content-connect:related` links to REST API responses for all post types that support REST, enabling discoverability of relationship endpoints.

All endpoints include proper permission checks, input validation, error handling, and integration tests covering authentication, pagination, filtering, ordering, and edge cases.

Closes #78

### How to test the Change

1. **V1 Search Endpoint:**
   - Make a POST request to `/content-connect/v1/search` with a valid nonce and search query
   - Test searching for posts and users separately
   - Verify pagination works correctly

2. **V2 Relationships Endpoint:**
   - GET `/content-connect/v2/relationships` to list all relationships
   - Test filtering by `rel_type`, `filter_by`, and `filter_value` parameters

3. **V2 Post Relationships Endpoint:**
   - GET `/content-connect/v2/post/{post_id}/relationships` to retrieve relationships for a specific post
   - Test filtering by `rel_type`, `post_type`, and `context` parameters
   - Verify `context=embed` includes related entities in the response

4. **V2 Post Related Entities Endpoint:**
   - GET `/content-connect/v2/post/{post_id}/related?rel_key={key}&rel_type={type}` to retrieve related entities
   - Test pagination, ordering, and filtering
   - POST to update the complete list of related entities
   - PUT to add a single related entity
   - DELETE to remove a related entity

5. **REST API Links:**
   - Retrieve any post via the WordPress REST API (e.g., `/wp/v2/posts/{id}`)
   - Verify that `_links` includes `content-connect:relationships` and `content-connect:related` links when the post has relationships

### Changelog Entry

> Added - `POST /content-connect/v1/search` - Search endpoint for posts and users with nonce-based authentication and pagination support.
> Added - `GET /content-connect/v2/relationships` - Retrieves all relationships by type with optional filtering.
> Added - `GET /content-connect/v2/post/{post_id}/relationships` - Retrieve all relationships associated with a post with filtering and context support.
> Added - `GET /content-connect/v2/post/{post_id}/related` - Retrieve entities related to a post with pagination, ordering, and filtering support.
> Added - `POST /content-connect/v2/post/{post_id}/related` - Update the complete list of related entities for a post.
> Added - `PUT /content-connect/v2/post/{post_id}/related` - Add a related entity to a post.
> Added - `DELETE /content-connect/v2/post/{post_id}/related` - Remove a related entity from a post.
> Added - Automatic REST API link integration for post types that support REST, adding `content-connect:relationships` and `content-connect:related` links to post responses.
> Added - Integration tests for all REST API endpoints covering authentication, permissions, pagination, filtering, ordering, and error cases.

### Credits

Props @s3rgiosan

### Checklist:

- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [x] I have updated the documentation accordingly.
- [x] I have added [Critical Flows, Test Cases, and/or End-to-End Tests](https://10up.github.io/Open-Source-Best-Practices/testing/) to cover my change.
- [x] All new and existing tests pass.

### CODE REVIEW

**Comment by s3rgiosan** [comment]
Hi @rickalee @jeffpaul, this is ready for review. This PR provides the foundational work for #94, and #97.

**Comment by s3rgiosan** [comment]
I've created a `try/block-editor-ui` all-in-one branch to make it easier to test the features in both the block and classic editors, as well as the Query Loop integration.

**Comment by s3rgiosan** [comment]
Added a new endpoint `GET /content-connect/v2/relationships` that retrieves all relationships by type (post-to-post or post-to-user).

File: includes/API/V2/Route/Relationships.php [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+	 */
+	public function get_items( \WP_REST_Request $request ) {
+
+		$rel_type = $request->get_param( 'rel_type' );
+
+		$relationships = array();
+
+		switch ( $rel_type ) {
+			case 'post-to-post':
+				$relationships = get_post_to_post_relationships_by();
```
Comment (benlk) [inline includes/API/V2/Route/Relationships.php]: It would be nice to be able to pass arguments to this function: `$field` and `$value` would let users filter this list to a specific post type.

**Comment by benlk** [comment]
The new API endpoint at `/content-connect/v2/relationships/` is returning 404 on my site, and doesn't appear in the enumerationn of endpoints at `/content-connect/v2/`.

File: includes/API/V2/Route/Relationships.php (hunk 1 above)
Comment (s3rgiosan) [reply includes/API/V2/Route/Relationships.php]: It's possible to filter relationships now that I added the `filter_by` and `filter_value` parameters to the endpoint, @benlk.

**Comment by s3rgiosan** [comment]
> The new API endpoint at `/content-connect/v2/relationships/` is returning 404 on my site, and doesn't appear in the enumerationn of endpoints at `/content-connect/v2/`.

@benlk Which branch are you using? This one or the `try/block-editor-ui` branch?

**Comment by s3rgiosan** [comment]
@rickalee I've added the links that we talked about on our last call. 

For a specific post, the WP core REST endpoint (`http://content-connect.test/wp-json/wp/v2/<post_type>/<post_id>`) response will be something like:

Screenshot 2025-05-26 at 18 25 15

**Comment by benlk** [comment]
> @benlk Which branch are you using? This one or the `try/block-editor-ui` branch?

I was using `try/block-editor-ui`, but I'm no longer on the project that was building off of this, so I won't be able to test this further.

**Comment by s3rgiosan** [comment]
@rickalee PR is ready for review. The `1.7.0-beta.1` tag on `release/1.7.0` should help with testing in your projects. Let me know if you find any issues.

cc @jeffpaul @fabiankaegy

**Review by jeffpaul — APPROVED** [review]
